extensions/guacamole-auth-radius/src/main/java/org/apache/guacamole/auth/radius/RadiusConnectionService.java

```diff
@@ -0,0 +1,352 @@
+        }
+
+        // Create the connection and load the attribute dictionary
+        createRadiusConnection();
+        AttributeFactory.loadAttributeDictionary("net.jradius.dictionary.AttributeDictionaryImpl");
+
+        // If the client is null, we return null - something has gone wrong
+        if (radiusClient == null)
+            return null;
+
+        RadiusAuthenticator radAuth = setupRadiusAuthenticator();
+
+        if (radAuth == null)
+            throw new GuacamoleException("Unknown RADIUS authentication protocol.");
+
+        // Set up attributes, create the access request, and send the packet
+        try { 
+            AttributeList radAttrs = new AttributeList();
+            radAttrs.add(new Attr_UserName(username));
+            radAttrs.add(new Attr_UserPassword(password));
+            radAttrs.add(new Attr_CleartextPassword(password));
+
+            AccessRequest radAcc = new AccessRequest(radiusClient);
+
+            // EAP-TTLS tunnels protected attributes inside the TLS layer
+            if (radAuth instanceof EAPTTLSAuthenticator) {
+                radAuth.setUsername(new Attr_UserName(username));
+                ((EAPTTLSAuthenticator)radAuth).setTunneledAttributes(radAttrs);
+            }
+            else
+                radAcc.addAttributes(radAttrs);
+
+            radAuth.setupRequest(radiusClient, radAcc);
+            radAuth.processRequest(radAcc);
+            RadiusResponse reply = radiusClient.sendReceive(radAcc, confService.getRadiusRetries());
+
+            // We receive a Challenge not asking for user input, so silently process the challenge
+            while((reply instanceof AccessChallenge) && (reply.findAttribute(Attr_ReplyMessage.TYPE) == null)) {
+                radAuth.processChallenge(radAcc, reply);
+                reply = radiusClient.sendReceive(radAcc, confService.getRadiusRetries());
+            }
+            return reply;
+        }
+
+        catch (RadiusException e) {
+            logger.error("Unable to complete authentication.", e.getMessage());
+            logger.debug("Authentication with RADIUS failed.", e);
+            return null;
+        }
+
+        catch (NoSuchAlgorithmException e) {
+            logger.error("No such RADIUS algorithm: {}", e.getMessage());
+            logger.debug("Unknown RADIUS algorithm.", e);
+            return null;
+        }
+    }
+
+    /**
+     * Authenticate to the RADIUS server using existing state and a response
+     *
+     * @param username
+     *     The username for the authentication
+     * @param state
+     *     The previous state of the RADIUS connection
+     * @param response
+     *     The response to the RADIUS challenge
+     *
+     * @return
+     *     A RadiusPacket with the response of the server.
+     *
+     * @throws GuacamoleException
+     *     If an error occurs while talking to the server.
+     */
+    public RadiusPacket authenticate(String username, String state, String response)
```

Comment:
Well, after looking at this, I did a little bit different swap on it - instead of making the response field optional, I made the state field optional, and only add that attribute to the RADIUS request packet when present.

I then made a sort of skeleton method called sendChallengeResponse that checks for the fields and then passes through to the authenticate method.  The only reason for this particular method is to help with debugging - it provides messages specific to issues with challenge/response authentication, but since the data all ends up in roughly the same place as it does with initial authentication, it can just pass it through.  This isn't necessary, per se - it functions perfectly fine with a single method - it's just to help with debugging issues specific to challenge/response that would result from missing the state or response.